src/BaseVault.sol

```diff
@@ -0,0 +1,662 @@
+    /**
+     * @notice Sets the buffer strategy.
+     * @param buffer_ The address of the buffer strategy.
+     */
+    function setBuffer(address buffer_) external virtual onlyRole(BUFFER_MANAGER_ROLE) {
+        if (buffer_ == address(0)) {
+            revert ZeroAddress();
+        }
+
+        _getVaultStorage().buffer = buffer_;
+        emit SetBuffer(buffer_);
+    }
+
+    /**
+     * @notice Sets the processor rule for a given contract address and function signature.
+     * @param target The address of the target contract.
+     * @param functionSig The function signature.
+     * @param rule The function rule.
+     */
+    function setProcessorRule(address target, bytes4 functionSig, FunctionRule calldata rule)
+        public
+        virtual
+        onlyRole(PROCESSOR_MANAGER_ROLE)
+    {
+        _getProcessorStorage().rules[target][functionSig] = rule;
+        emit SetProcessorRule(target, functionSig, rule);
+    }
+
+    /**
+     * @notice Adds a new asset to the vault.
+     * @param asset_ The address of the asset.
+     * @param active_ Whether the asset is active or not.
+     */
+    function addAsset(address asset_, bool active_) public virtual onlyRole(ASSET_MANAGER_ROLE) {
+        _addAsset(asset_, IERC20Metadata(asset_).decimals(), active_);
+    }
+
+    function _addAsset(address asset_, uint8 decimals_, bool active_) internal virtual {
+        if (asset_ == address(0)) {
+            revert ZeroAddress();
+        }
+        AssetStorage storage assetStorage = _getAssetStorage();
+        uint256 index = assetStorage.list.length;
+        if (index > 0 && assetStorage.assets[asset_].index != 0) {
+            revert DuplicateAsset(asset_);
+        }
+        assetStorage.assets[asset_] = AssetParams({active: active_, index: index, decimals: decimals_});
+        assetStorage.list.push(asset_);
+
+        emit NewAsset(asset_, decimals_, index);
+    }
+
+    /**
+     * @notice Pauses the vault.
+     */
+    function pause() external virtual onlyRole(PAUSER_ROLE) {
+        if (paused()) {
+            revert Paused();
+        }
+
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        vaultStorage.paused = true;
+        emit Pause(true);
+    }
+
+    /**
+     * @notice Unpauses the vault.
+     */
+    function unpause() external virtual onlyRole(UNPAUSER_ROLE) {
+        if (!paused()) {
+            revert Unpaused();
+        }
+
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        if (provider() == address(0)) {
+            revert ProviderNotSet();
+        }
+        vaultStorage.paused = false;
+        emit Pause(false);
+    }
+
+    /**
+     * @notice Processes the accounting of the vault by calculating the total base balance.
+     * @dev This function iterates through the list of assets, gets their balances and rates,
+     *      and updates the total assets denominated in the base asset.
+     */
+    function processAccounting() public virtual {
```

Comment:
also on `processAccounting` -- when are we planning to call this? and should we continuously call this for every vault/asset/strategy (not sure what's the right terminology here) -- is it realistic both from a gas perspective and user security perspective?